lib/Dialect/SV/Transforms/RTLLegalizeNames.cpp

```diff
@@ -0,0 +1,198 @@
+  /// keywords.
+  StringRef getModuleName(StringAttr moduleName) {
+    auto entryIter = moduleNameTable.find(moduleName);
+    if (entryIter != moduleNameTable.end())
+      return entryIter->getSecond();
+    auto updatedName = sanitizeName(moduleName.getValue(), usedModuleNames,
+                                    nextGeneratedNameID);
+    moduleNameTable[moduleName] = updatedName;
+    return updatedName;
+  }
+
+  /// moduleNameTable keeps track of mappings from Module names to updated names
+  /// to resolve keyword conflicts.
+  llvm::DenseMap<Attribute, StringRef> moduleNameTable;
+
+  /// Set of used MLIR module names, to ensure unique names when renaming
+  /// keywords in module names.
+  llvm::StringSet<> usedModuleNames;
+
+  size_t nextGeneratedNameID = 0;
+};
+} // end anonymous namespace
+
+void RTLLegalizeNamesPass::runOnOperation() {
+  anythingChanged = false;
+  auto root = getOperation();
+
+  // We run in two phases. Phase one legalizes module and interface names in the
+  // module globally. Phase two then dives into these structures and performs
+  // local legalization on names that cannot leak out their enclosing module.
+
+  // Register the names of external modules which we cannot rename.
+  root.getBodyRegion().walk([&](Operation *op) {
+    if (auto extMod = dyn_cast<RTLModuleExternOp>(op)) {
+      auto verilogName = extMod.getVerilogModuleNameAttr();
+      getModuleName(verilogName);
+      // We expressely *do not* rename extern modules, where the user supplies
+      // the complementary definition from the outside.
+      // TODO: Maybe we should emit a diagnostic if this pass would rename an
+      // external module.
+    }
+  });
+
+  // Rename modules and interfaces.
+  mlir::SymbolTableCollection symbolTable;
+  mlir::SymbolUserMap symbolUsers(symbolTable, root);
+  root.getBodyRegion().walk([&](Operation *op) {
+    StringAttr nameAttr;
+    if (auto module = dyn_cast<RTLModuleOp>(op)) {
+      nameAttr = module.getNameAttr();
+    } else if (auto intf = dyn_cast<InterfaceOp>(op)) {
+      nameAttr = intf.getNameAttr();
+    }
+    if (nameAttr) {
+      auto oldName = nameAttr.getValue();
+      auto newName = getModuleName(nameAttr);
+      if (oldName != newName) {
+        symbolUsers.replaceAllUsesWith(op, newName);
+        SymbolTable::setSymbolName(op, newName);
+      }
+      anythingChanged = true;
+    }
+  });
+
+  // Tables to track the used names within a given module or interface.
+  // llvm::DenseMap<ValueOrOp, StringRef> nameTable;
+  llvm::StringSet<> usedNames = usedModuleNames;
+  size_t nextGeneratedNameID = 0;
+
+  // Rename individual operations.
+  root.getBodyRegion().walk([&](Operation *op) {
+    if (auto module = dyn_cast<RTLModuleOp>(op)) {
+      auto moduleType = rtl::getModuleType(module);
+      auto inputs = moduleType.getInputs();
+      auto results = moduleType.getResults();
+
+      // Rename the inputs.
```

Comment:
I don't think there's an issue with port names conflicting with any other names. They just can't be keywords or conflict with themselves.